lib/spack/spack/compilers/msvc.py

```diff
@@ -164,7 +164,7 @@ def setup_custom_environment(self, pkg, env):
             out = out.decode("utf-16le", errors="replace")  # novermin
 
         int_env = dict(
-            (key.lower(), value)
+            (key.upper(), value)
```

Comment:
Is there a reason not to use the case from setvars? does it not match the needed case for the environment variables? I ask in case some variable is case-sensitive and not all-caps.

Reply:
Windows sets the path env var to be `Path` not `PATH`. It does this with a number of other env vars as well, preferring pascal case (which is also the .NET convention). This causes issues with Spack env settings as while Spack's env module makes distinctions along case lines, the Windows env does not disambiguate along those lines. So Spack setting `Path` and `PATH` means whichever is set first will be eclipsed by the latter. In general, from a grep of the codebase, Spack seems to prefer the uppercase option, so this is meant to conform to that convention on the python side because the Windows side doesn't care. This behavior on Windows is a holdover from the original DOS implementations.

Reply:
Based on this, it sounds like it would make sense for `EnvironmentModifications` to handle this (so that we don't accidentally introduce differently-cased instances of e.g. PATH in other places): that is, on Windows the functions like `append_path` etc. always convert `name` to be upper case.

Reply:
So, make all the methods on `EnvironmentModifications` call upper on Windows for the env var?  That might prevent bugs like this in the future or even fix some we don't know about. Might make `EnvironmentModifications`  methods look a bit ugly `name = self.platform_normalize(name)` where there could be an if windows in platform_normalize otherwise it returns just name.

Reply:
In general when we're applying what is essentially a filter operation on method arguments, I prefer to use decorators for this reason as they're fairly low profile in terms of readability/code flow.